Implement the Arm

jordan-powers opened the pull request:
This PR implements the ArmSubsystem and TeleopArmCommand per Jackson's description (which is included below).

Much of this code should be familiar, as it is very similar to the arm control code from 2023.

>Arm:
2 NEOs - Arm Shoulder Motors - These are two motors that are connected to a chain and sprocket system that drives the shoulder joint. The maximum expected travel of the arm should never exceed a 165 degree arc at the maximum. One of these motors will have a REV Through Bore Encoder for an absolute position value, Also, these motors will be rotating opposite to each other. There are five positions that the arm can be in: stowed, handoff, shooting, amp scoring, and a back up position to intake directly from the Source if the ground intake takes a major hit and is knocked out of commission for a match. A six possible position may be needed if we attempt to score in the trap.
I think the set point system from 2023 would work fine here. A few changes like the handoff button and the shooting button controlling the set point might be necessary. Also, I think the arm so default to being stowed when a button is not held down so we can move under the Stage without concern. Software limits on when set points can be activated would be nice (like a bounding box around the stage to prevent an arm extension from underneath). Software limits on rotation would also be nice, I imagine have backup manual control like we did for 2023 would be useful.
1 NEO - Arm Wrist Motor - This a single motor that drives the wrist joint. The wrist has a REV Through Bore Encoder for absolute position. This joint needs to be coordinated with the other set points of the arm but also when shooting, this joint should be doing the aiming.
Once again, I think the wrist should operate just like the arm using setpoints with the only difference between the 2023 bot is that it can auto aim in the shooting configuration.

jordan-powers committed:
Add ArmSubsystem

jordan-powers committed:
Add ArmSetpointManager using persistent NetworkTable entries

jordan-powers committed:
Add TeleopArmCommand

jordan-powers committed:
Add DualControllerInput

jordan-powers committed:
Fix setPersistent in ArmSetpointManager::DataStore

Turns out NetworkTables ignores setPersistent() if there's no value yet
published to the topic. The solution is to use setDefault() to publish
the default value before calling setPersistent().

jordan-powers committed:
Instantiate DualControllerInput class in RobotContainer

jordan-powers committed:
Set TeleopArmCommand as the default command for the ArmSubsystem

dslit3 commented on src/main/java/frc/robot/util/MoPrefs.java:
```diff
@@ -56,6 +56,26 @@ public class MoPrefs {
+    public static final UnitPref<Velocity<Angle>> wristMaxRps =
+            rotationsPerSecPref("Wrist Max Speed", Units.RotationsPerSecond.of(0.5));
+
+    public static final Pref<Double> armWrampTime = unitlessDoublePref("Arm Ramp Time", 0.15);
```
`armWrampTime` -> `armRampTime`

jordan-powers replied:
Oops, good catch!

jordan-powers committed:
Use ArmFeedforward to linearize shoulder response

jordan-powers committed:
Add ArmControlMode chooser to shuffleboard

jordan-powers committed:
Setup sysid run

jordan-powers committed:
Add command to aim the arm/wrist at the speaker

jordan-powers committed:
Use MoEncoder in ArmSubsystem

jordan-powers committed:
Merge branch 'main' into arm

jordan-powers committed:
Merge branch 'main' into arm

jordan-powers committed:
Add arm control to JoystickInput class

jordan-powers committed:
Set theoretical shoulder encoder ratios as defaults

momentumrobotics-4999 committed:
Add shoulder CAN ids

momentumrobotics-4999 committed:
Update shuffleboard label positions

momentumrobotics-4999 committed:
Reverse shoulder motors

momentumrobotics-4999 committed:
Enable soft limits on motor controller

jordan-powers committed:
Add button to re-zero arm during match

jordan-powers committed:
Use TunerUtils.forSparkMaxArm for shoulder controllers

jordan-powers committed:
Add title to arm control mode chooser

jordan-powers committed:
Update sysid config parameters

jordan-powers committed:
Fix sysid get voltage from sparkmax

jordan-powers committed:
MoSparkMaxArmPID actually save kS,kG,kV params when set

jordan-powers committed:
MoSparkMaxArmPID specify feedforward units

jordan-powers committed:
Remove unit conversion from MoSparkMaxPID

To keep things organized, only the MoEncoder class should know the internal
units of the encoder. Instead of passing the unit itself into
MoSparkMaxPID, instead we just pass the MoEncoder that the unit lives
within.

jordan-powers committed:
Add latest pid constants and robot preferences

jordan-powers committed:
Add shoulder/wrist velocities to shuffleboard

momentumrobotics-4999 committed:
Add sysid constants to shuffleboard

momentumrobotics-4999 committed:
Fix kV is in volts seconds per rotation

jordan-powers committed:
Add TeleopDriveWithPointAtCommand

jordan-powers committed:
Reset didEstablishInitialPosition on alliance change

jordan-powers committed:
Reverse joystick twist axis

jordan-powers committed:
Fix documentation on Limelight::isPoseDataValid

jordan-powers committed:
Remove limelightTransform

The limelight's offset should be handled by the limelight internally.
